crates/aptos-speculative/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+// Suppress incorrect warning, as without mut the take_counts method stops compiling.
+#![allow(unused_mut)]
+
+use aptos_infallible::Mutex;
+use crossbeam::utils::CachePadded;
+use once_cell::sync::OnceCell;
+use rayon::prelude::*;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[cfg(test)]
+mod tests;
+
+/// This crate is designed to help with tracking events and state that might be speculative in
```

Comment:
Can you add one specific example how the log can be used, i.e., for speculative execution in parallel execution.

Reply:
referred to tests/logging.rs and made it also serve as an example ;-)